### PR TITLE
sasjs init

### LINKED ISSUE #479: sasjs init
We need a command for initialising a SASjs repo inside an existing repository, this will create:

 - sasjs folder (including doxy and sasjsconfig.json)
 - package.json (create if not exists, else update with sasjs/cli and sasjs/core dependency)
 - .gitignore (node_modules & sasjsbuild if not already listed)

### PR DESCRIPTION
**Pull request by saadjutt01**
## Issue

Closes #479 

## Intent

Add new command 'sasjs init'

## Implementation

- added `init.ts`
- added `./shared/createConfigFile.ts `
- extracted code from `create.ts` to `./shared/createFileStructure.ts`
- updated `setupGitIgnore()` with `node_modules`

## Checks

- [x] Code is formatted correctly (`npm run lint:fix`).
- [x] Any new functionality has been unit tested.
- [x] All unit tests are passing (`npm test`).
- [ ] All CI checks are green.
- [x] JSDoc comments have been added or updated.
- [x] Reviewer is assigned.

### CODE REVIEW

File: src/main.ts [hunk 1]
```diff
@@ -30,6 +31,19 @@ export enum ReturnCode {
   InternalError
 }
 
+export async function initSasjs() {
+  return await init()
+    .then(() => {
+      displaySuccess(
+        'Project is powered with SASjs. You can use any sasjs command, now.'
```
Comment (allanbowe) [inline src/main.ts:38]: nice!

**Review by allanbowe — APPROVED** [review]

**Review by allanbowe — APPROVED** [review]

File: src/commands/init/init.ts [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+  setupNpmProject,
+  setupGitIgnore,
+  setupDoxygen
+} from '../../utils/utils'
+
+import { createFolder } from '../../utils/file'
+import { createConfigFile } from '../shared/createConfigFile'
+
+export async function init() {
+  process.logger?.info('Creating folders and files...')
```
Comment (krishna-acondy) [inline src/commands/init/init.ts:13]: It might be good to have a different message here than what we have for `sasjs create`.

File: src/main.ts [hunk 3]
```diff
@@ -32,6 +33,19 @@ export enum ReturnCode {
...
```
Comment (krishna-acondy) [inline src/main.ts:40]: I'd suggest something along the lines of - `This project is powered by SASjs. You can now use any sasjs command within the project. For more information, type 'sasjs help' or visit https://cli.sasjs.io/`

@allanbowe what do you think?

File: src/commands/init/init.ts (hunk 2 above)
Comment (saadjutt01) [reply src/commands/init/init.ts:13]: What should be better message on cli, @allanbowe 

> `Setting up environment for SASjs`
> `Placing minimum required files for SASjs`
> `Making it ready for SASjs`

;)

Comment (allanbowe) [reply src/commands/init/init.ts:13]: How about `Initialising SASjs...`

**Review by allanbowe — APPROVED** [review]

File: src/commands/init/init.ts (hunk 2 above)
Comment (allanbowe) [reply src/commands/init/init.ts]: fixed

**Review by allanbowe — APPROVED** [review]

File: src/commands/help/help.ts [hunk 4]
```diff
@@ -2,6 +2,15 @@ import chalk from 'chalk'
 
 export async function printHelpText() {
   const commands = [
+    {
+      name: 'init',
+      title: 'init',
+      description: [
+        `creates the folder 'sasjs' and place doxy folder (having content files + configuration) for docs`,
```
Comment (YuryShkoda) [inline src/commands/help/help.ts:9]: `place` -> `places`

File: src/commands/init/spec/init.spec.server.ts [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    dotenv.config()
+  })
+
+  afterEach(async (done) => {
+    await deleteFolder(path.join(__dirname, 'test-app-init-*'))
+    done()
+  })
+
+  it('should initialise with default app in the current folder', async (done) => {
+    const appName = `test-app-init-.-${generateTimestamp()}`
```
Comment (YuryShkoda) [inline src/commands/init/spec/init.spec.server.ts:21]: we usually name test apps `test-app-init-${generateTimestamp()}`

File: src/commands/init/spec/init.spec.server.ts [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+    await expect(init()).toResolve()
+
+    await verifyInit()
+
+    done()
+  })
+
+  it(`should initialise with 'sasonly' app in the current folder`, async (done) => {
+    const appName = `test-app-init-.-${generateTimestamp()}-sasonly`
```
Comment (YuryShkoda) [inline src/commands/init/spec/init.spec.server.ts:37]: `test-app-init-.-${generateTimestamp()}-sasonly` -> `test-app-init-sasonly-${generateTimestamp()}`

File: src/commands/shared/createFileStructure.ts [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+import { asyncForEach } from '../../utils/utils'
+import { getFolders } from '../../utils/config'
+import { createFolderStructure, fileExists } from '../../utils/file'
+import { Folder } from '../../types'
+
+import { createConfigFile } from './createConfigFile'
+
+/**
+ * Creates the folder structure specified in config.json
+ * Also Creates a SASjs configation file, named 'sasjsconfig.json'.
```
Comment (YuryShkoda) [inline src/commands/shared/createFileStructure.ts:12]: `Creates` -> `creates`

File: src/utils/utils.ts [hunk 8]
```diff
@@ -126,14 +126,19 @@ export async function setupGitIgnore(folderPath: string): Promise<void> {
   if (gitIgnoreExists) {
     const gitIgnoreContent = await readFile(gitIgnoreFilePath)
 
-    const regExp = new RegExp(`^sasjsbuild\/`, 'gm')
+    const regExp1 = new RegExp(`^sasjsbuild\/`, 'gm')
```
Comment (YuryShkoda) [inline src/utils/utils.ts:129]: Please improve variable naming

File: src/utils/utils.ts [hunk 9]
```diff
@@ -126,14 +126,19 @@ export async function setupGitIgnore(folderPath: string): Promise<void> {
...
+    const regExp2 = new RegExp(`^node_modules\/`, 'gm')
```
Comment (YuryShkoda) [inline src/utils/utils.ts:130]: Please improve variable naming

File: src/main.ts [hunk 10]
```diff
@@ -32,6 +33,19 @@ export enum ReturnCode {
...
+export async function initSasjs() {
+  return await init()
+    .then(() => {
+      displaySuccess(
+        'This project is powered by SASjs. You can now use any sasjs command within the project. For more information, type "sasjs help" or visit https://cli.sasjs.io/'
+      )
+      return ReturnCode.Success
+    })
+    .catch((err: any) => {
+      displayError(err, 'An error has occurred whilst initiating sasjs init.')
```
Comment (YuryShkoda) [inline src/main.ts:45]: `whilst` -> `while`

**Review by allanbowe — APPROVED** [review]

**Pull request merged**